sklearn_numba_dpex/kmeans/engine.py

```diff
@@ -47,101 +63,134 @@ class KMeansEngine(KMeansCythonEngine):
-            )
+        return X_t.T, y, self.sample_weight
 
-        X, sample_weight, _, output_dtype = self._check_inputs(
-            X, self.sample_weight, cluster_centers=None
-        )
+    def unshift_centers(self, X, best_centers):
+        if (X_mean := self.X_mean) is None:
+            return
 
-        X_t, sample_weight, _ = self._load_transposed_data_to_device(
-            X, sample_weight, cluster_centers=None
-        )
+        best_centers += dpt.asnumpy(X_mean.get_array())
 
-        centers, center_indices = kmeans_plusplus(
-            X_t, sample_weight, n_clusters, self.random_state
-        )
+        # NB: self.estimator.copy_x being set to False does not mean that no copy
+        # actually happened, only that no copy was forced if it was not necessary
+        # with respect to what device, dtype and order that are required at compute
+        # time. Nevertheless, there's no simple way to check if a copy happened
+        # without assumptions on the type of the raw input submitted by the user,
+        # but at the moment it is unknown what those assumptions could be.
+        # As a result, the following instructions are ran every time, even if it
+        # isn't useful when a copy has been made.
+        # TODO: is there a set of assumptions that exhaustively describes the set
+        # of accepted inputs, and also enables checking if a copy happened or not
+        # in a simple way ?
+        if not self.estimator.copy_x:
+            restore_data_after_lloyd(X.T, X_mean)
+
+    def init_centroids(self, X):
+        init = self.init
+        n_clusters = self.estimator.n_clusters
 
-        centers = dpt.asnumpy(centers).astype(output_dtype, copy=False)
-        center_indices = dpt.asnumpy(center_indices)
-        return centers, center_indices
+        if isinstance(init, dpt.usm_ndarray):
+            centers_t = init
 
-    def kmeans_single(self, X, sample_weight, centers_init):
-        X, sample_weight, cluster_centers, output_dtype = self._check_inputs(
-            X, sample_weight, centers_init
-        )
+        elif isinstance(init, str) and init == "k-means++":
+            centers_t, _ = self._kmeans_plusplus(X)
+
+        elif callable(init):
+            centers = init(X, self.estimator.n_clusters, random_state=self.random_state)
+            centers_t = self._check_init(centers, X)
+
+        else:
+            # NB: sampling without replacement must be executed sequentially so
+            # it's better done on CPU
+            centers_idx = self.random_state.choice(
+                X.shape[0], size=n_clusters, replace=False
+            )
+            # Poor man's fancy indexing
+            # TODO: write a kernel ? or replace with better equivalent when available ?
+            centers_t = dpt.concat(
+                [dpt.expand_dims(X[center_idx], axes=1) for center_idx in centers_idx],
+                axis=1,
+            )
```

Comment:
Should we open an issue on `dpctl`? What do you think?

Reply:
`dpctl` implements the Array API and the Array API doesn't list such tools. Best things to have would be `take` but here are the list of supported functions [1](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html) [2](https://data-apis.org/array-api/latest/API_specification/indexing.html)

I'm not aware enough of where the array api is going to say what would be the best action here :thinking:

Reply:
Actually `take` has been added recently to the Array APi, [dpctl issue](https://github.com/IntelPython/dpctl/issues/1003)